pkg/kubelet/cm/node_allocatable.go

```diff
@@ -0,0 +1,234 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cm
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/kubelet/events"
+	evictionapi "k8s.io/kubernetes/pkg/kubelet/eviction/api"
+)
+
+type nodeAllocatableCalculationPurpose int
+
+const (
+	defaultNodeAllocatableCgroupName = "kubepods"
+	NodeAllocatableEnforcementKey    = "pods"
+	SystemReservedEnforcementKey     = "system-reserved"
+	KubeReservedEnforcementKey       = "kube-reserved"
+	// Constants for identifying various types of Node Allocatable reservations.
+	forScheduling  nodeAllocatableCalculationPurpose = iota // 0
+	forEnforcement nodeAllocatableCalculationPurpose = iota
+)
+
+func (cm *containerManagerImpl) createNodeAllocatableCgroups() error {
+	cgroupConfig := &CgroupConfig{
+		Name: CgroupName(cm.cgroupRoot),
+	}
+	if cm.cgroupManager.Exists(cgroupConfig.Name) {
+		return nil
+	}
+	if err := cm.cgroupManager.Create(cgroupConfig); err != nil {
+		glog.Errorf("Failed to create %q cgroup", cm.cgroupRoot)
+		return err
+	}
+	return nil
+}
+
+// runUntilSuccess runs `f` until it returns nil.
+// It invokes `success` func on success and returns.
+// It invokes `failure` func on failure and retries after `retryPeriod` duration.
+func runUntilSuccess(f func() error, success func(), failure func(error), retryPeriod time.Duration) {
+	for {
+		err := f()
+		if err == nil {
+			success()
+			return
+		}
+		failure(err)
+		time.Sleep(retryPeriod)
+	}
+}
+
+// Enforce Node Allocatable Cgroup settings.
+func (cm *containerManagerImpl) enforceNodeAllocatableCgroups() error {
+	nc := cm.NodeConfig.NodeAllocatableConfig
+	nodeAllocatable := cm.getNodeAllocatableInternal(forEnforcement)
+
+	glog.V(4).Infof("Attempting to enforce Node Allocatable with config: %+v", nc)
+	glog.V(4).Infof("Node Allocatable resources: %+v", nodeAllocatable)
+	// Create top level cgroups for all pods if necessary.
+	if nc.EnforceNodeAllocatable.Has(NodeAllocatableEnforcementKey) {
+		cgroupConfig := &CgroupConfig{
+			Name:               CgroupName(cm.cgroupRoot),
+			ResourceParameters: getCgroupConfig(nodeAllocatable),
+		}
+		// If Node Allocatable is enforced on a node that has not been drained or is updated on an existing node to a lower value,
+		// existing memory usage across pods might be higher that current Node Allocatable Memory Limits.
+		// Pod Evictions are expected to bring down memory usage to below Node Allocatable limits.
+		// Until evictions happen retry cgroup updates.
+		go runUntilSuccess(
```

Comment:
That's possible, but that would result in a larger anonymous method just like the previous iteration.